src/MongoDB.Bson/Serialization/Serializers/ObjectSerializer.cs

```diff
@@ -88,12 +102,28 @@ public override object Deserialize(BsonDeserializationContext context, BsonDeser
                     goto default;
 
                 case BsonType.Binary:
-                    var binaryData = bsonReader.ReadBinaryData();
-                    var subType = binaryData.SubType;
-                    if (subType == BsonBinarySubType.UuidStandard || subType == BsonBinarySubType.UuidLegacy)
+#pragma warning disable 618
+                    if (BsonDefaults.GuidRepresentationMode == GuidRepresentationMode.V2 && _guidRepresentation == GuidRepresentation.Unspecified)
                     {
-                        return binaryData.ToGuid();
+                        var binaryData = bsonReader.ReadBinaryData();
+                        var subType = binaryData.SubType;
+                        if (subType == BsonBinarySubType.UuidStandard || subType == BsonBinarySubType.UuidLegacy)
+                        {
+                            return binaryData.ToGuid();
+                        }
+                    }
+                    else
+                    {
+                        var binaryDataBookMark = bsonReader.GetBookmark();
+                        var binaryData = bsonReader.ReadBinaryDataIgnoringGuidRepresentation();
```

Comment:
We need to use the new ReadBinaryDataIgnoringGuidRepresentation method to suppress spurious exceptions from ReadBinaryData that don't apply here.